### ISSUE TITLE
PRN module training collapse

### ISSUE THREAD

**Issue by lDNAl** [firework8/ProtoGCN#18]
Hi,

Thanks for sharing your code and great work! 

I am replicating the experiment with the configuration for ntu120-xsub-j. With the same configuration with 4gpu(video_per_gpu=16,lr=0.025) or 1gpu(video_per_gpu=64,lr=0.1), the PRN module is easy to fail training which gives constants for reconstructions after the PRN module or the reconstruction refinement(post->bn->relu). The pretrained models given by your team is outputting distinct reconstructions. Did you have the same training issue when you trained for that many different experiment setups? If so, how did you avoid it?

Thanks for your time and reading.

**Comment by firework8** [comment]
Thank you for your interest in our work.

I’m not entirely sure what you mean by training collapse. If it refers to high loss values, you can refer to [issues#12](https://github.com/firework8/ProtoGCN/issues/12).

The current public settings are  actually the same as the pre-training settings. If performance fluctuates, it may be due to the randomness inherent in data augmentation. I recommend running the model a few more times with the same settings.

In addition, these two configurations are different. The setting for the 1-GPU version can be (video_per_gpu=16, lr=0.025) or (video_per_gpu=64, lr=0.1). But, the correct setting for the 4-GPU version should be (video_per_gpu=16, lr=0.1). It means 4 gpu*16=64, and maintains a ratio of 64 to 0.1.

I hope this helps. If any questions remain, please feel free to contact me.

**Comment by lDNAl** [comment]
Thanks for your reply!

Yes, you are right. I got a typo there. I mean I tried configurations with 4gpu(video_per_gpu=16,lr=0.1) and 1gpu(video_per_gpu=64,lr=0.1 or 0.025). 
After the training with above configurations, while the overall classification performance is close to the performance noted in the paper, the reconstruction from PRN module after post->bn->relu is giving the same constant for all elements which I regard as the training failure for PRN module.
My other concern is that when I train the model across 4 gpus, the code is not sharing the memory bank across gpus I believe. At the same time, the batch size would be 16 for each gpu, which should bring a difference than the given batch size as 64 for 1 gpu training. Do you have the same concern when you had the training with different amounts of used gpus?

**Comment by firework8** [comment]
Ok, I understand your concern.

1. I did visualize the reconstructed graph before and after, but I didn’t check the parameter values. I think your feedback actually offers a new perspective, which means the PRN module essentially serves as a form of motion feature compression, or mapping. I'll think about it some more later.

2. That should be fine, as the PYSKL framework supports parameter-parallel settings. In other words, the memory parameters can be optimized alongside those of other modules without being affected by the batch size or the number of GPUs. You can refer to [issues#168](https://github.com/kennymckormick/pyskl/issues/168#issuecomment-1518910511). Based on the results of my own experiments, as long as the same ratio is maintained, the number of GPUs used does not affect performance.

Thank you for your feedback!
Hope this helps.

**Comment by lDNAl** [comment]
1. The PRN training failing case only happens in my replication. The tricky part is that even when the PRN module is not trained well, the overall classification performance would still be quite close to the model trained without PRN module. As I said, I checked the reconstructions using your pretrained model(ntu120-xsub-j), and the reconstructions are successfully clustered and regularized by CSC module. I don't think it is a magical random seed case since I am not seeing one used in the code. It bothers me a lot since I am using ProtoGCN as the feature extractor these days.
2. Thanks for your explanation and ablation experiment information. It really helps!

**Comment by lDNAl** [comment]
I may have found the problem. In your paper, the input to the PRN module is A_inter^(L)+A_inter^(L)+A_0. But in the repo [line](https://github.com/firework8/ProtoGCN/blob/ddf7f274f9f5d9e45a2fcfeb299bfb3fd7c2303d/protogcn/models/gcns/utils/gcn.py#L113), there is only A_inter^(L)+A_inter^(L). As they are 0 set by alpha and beta at epoch 0, the PRN module is stuck in the uniform attention from softmax. Would you mind confirming if the code is right in your repo?

**Comment by firework8** [comment]
Sorry for the late reply.

Regarding point 1 mentioned above, PRN training is indeed somewhat challenging, as evidenced by the different parameters I set for distinct modalities. However, I can guarantee that the published pre-trained model was trained using the current configuration. I suggest running additional experiments with other modalities, such as ntu120-xsub-b.

For explanations regarding the MTE module, you can refer to [issues#4](https://github.com/firework8/ProtoGCN/issues/4). Additionally, if you are interested in obtaining better models and skeletal features, I recommend referring to my latest work [SkeletonAgent](https://github.com/firework8/SkeletonAgent), which can offer more stable performance.
